src/Spectre.libHeatmapDataScaling/HistogramEqualization.cpp

```diff
@@ -14,43 +14,55 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+#include "Spectre.libException/ArgumentOutOfRangeException.h"
 #include "Spectre.libFunctional/Transform.h"
 #include "Spectre.libHeatmapDataScaling/HistogramEqualization.h"
 
 namespace spectre::visualization
 {
-std::vector<double> HistogramEqualization::scaleData(const gsl::span<const double> intensities)
+std::vector<double> HistogramEqualization::scaleData(const gsl::span<const double> intensities) const
 {
-    std::vector<double> roundedIntensities = spectre::core::functional::transform(intensities, [](double intensity) { return static_cast<int>(intensity + 0.5); });
-    std::vector<uint8_t> intensitiesOccurances = countRepeatingValues(roundedIntensities);
-    std::vector<uint8_t> cumulativeDistribution = std::accumulate(intensitiesOccurances.begin(), intensitiesOccurances.end(), cumulativeDistribution, [](std::vector<uint8_t> accumulator, uint8_t intensityOccurences)
+	const auto minIntensitiesRange = *min_element(std::begin(intensities), std::end(intensities));
+	const auto maxIntensitiesRange = *max_element(std::begin(intensities), std::end(intensities));
+	if (minIntensitiesRange < 0)
+	{
+		throw core::exception::ArgumentOutOfRangeException<double>("minIntensitiesRange", 0, 255, minIntensitiesRange);
+	}
+	if (maxIntensitiesRange > 255)
+	{
+		throw core::exception::ArgumentOutOfRangeException<double>("maxIntensitiesRange", 0, 255, maxIntensitiesRange);
+	}
+
+	std::vector<uint8_t> roundedIntensities = spectre::core::functional::transform<double, uint8_t>(intensities, [](double intensity) { return static_cast<uint8_t>(intensity + 0.5); });
+    std::vector<unsigned int> intensitiesOccurences = countRepeatingValues(roundedIntensities);
+	const std::vector<unsigned int> cumulativeDistribution = std::accumulate(intensitiesOccurences.begin(), intensitiesOccurences.end(), std::vector<unsigned int>(), [](std::vector<unsigned int> accumulator, unsigned int intensityOccurences)
```

Comment:
You are copying `accumulator` in each call to the lambda. This is performance issue.

I digged a bit. It appears that [`std::partial_sum`](http://en.cppreference.com/w/cpp/algorithm/partial_sum) will be **much** more fitting this purpose than `std::accumulate`. It will be a pleasant one-liner, without this error-prone lambda below. 😃